modules/calib3d/src/solvepnp.cpp

```diff
@@ -309,6 +309,17 @@ bool solvePnPRansac(InputArray _opoints, InputArray _ipoints,
     int npoints1 = compressElems(&opoints_inliers[0], mask, 1, npoints);
     compressElems(&ipoints_inliers[0], mask, 1, npoints);
 
+    if ( flags == SOLVEPNP_ITERATIVE && npoints1 < 6 )
```

Comment:
I believe it is a little bit more complex.

When 3D points are not planar, `cvFindExtrinsicCameraParams2` uses DLT algorithm which indeed requires at least 6 points.

But with coplanar 3D points, `cvFindExtrinsicCameraParams2` should work with at least 4 points.

---

BTW, see here: https://github.com/opencv/opencv/issues/8782 my previous comment about the behavior of `solvePnPRansac` and this PR: https://github.com/opencv/opencv/pull/9086

In my opinion, behavior of `solvePnPRansac` is still not totally clear / intuitive.

Reply:
Ok, then `cvFindExtrinsicCameraParams2` may get good result with 4 points, but may throw Exception with 5 points.

I think CV_CheckGE in [`cvFindExtrinsicCameraParams2`](https://github.com/opencv/opencv/blob/863ecded30dcaedff0ece893277411f39bb3c1c6/modules/calib3d/src/calibration.cpp#L1099) is not a good idea. But since it returns void (not bool), throwing Exception may be the only option to inform that the result is bad.

As a result, try-catch seems inevitable.

Reply:
> Ok, then `cvFindExtrinsicCameraParams2` may get good result with 4 points, but may throw Exception with 5 points.

Not sure to follow correctly. In `cvFindExtrinsicCameraParams2` it checks first if the 3D points are coplanar. For coplanar case, it should never throw exception, except theoretically for degenerate cases like all collinear points but this is not handled/detected.

And there are already a check for at least 4 points if `useExtrinsicGuess` is false: https://github.com/opencv/opencv/blob/863ecded30dcaedff0ece893277411f39bb3c1c6/modules/calib3d/src/calibration.cpp#L1014

Reply:
`cvFindExtrinsicCameraParams2` does check if the 3D points are coplanar. And if not, it may throw exception.

When some one calls `cvFindExtrinsicCameraParams2` directly, he can make sure the input points are more than 4. If he knows that the points may not be coplanar, he can make sure the points are more than 6. In this cases, `cvFindExtrinsicCameraParams2` won't throw exceptions, because the user calls it in the "right" way.

However, codes in `solvePnPRansac` does not guarantee that. When the user knows the points used to solve PnP are very likely to be noncoplanar, the only thing he can do is to input as many points as possible (6 at least). But RANSACE procedure (the intermediate calculation) may reduce the points and make them not suitable for `cvFindExtrinsicCameraParams2`. That's why I think Exception here is not supposed to happen. It's NOT the user's fault.

To fix this, `solvePnPRansac` should either check the coplanarity before calling `solvePnP`, or call `solvePnP` with try-catch. The former solution may add unnecessary calculation.